### ISSUE TITLE
anzu--state is global

### ISSUE THREAD

**Issue by blaenk** [syohex/emacs-anzu#58]
I'm new to emacs so I'm not entirely sure, but I'm configuring the mode-line myself instead of using the default consing mechanism. So I set `anzu-mode-line-update-function` to my own function and `anzu-cons-mode-line-p` to nil. It works well. The only (major) problem is that if I search in one buffer, all of them show the same information with respect to anzu, in the mode-line.

I believe this is because the `anzu--state` variable is global, which is what I use to determine if the anzu stuff should be shown in my mode-line. It happens to work out fine in the default mechanism because it's only checked once to determine whether or not to cons the anzu stuff to the buffer-local mode-line. However, if you don't use a cons mechanism or otherwise update the mode-line once in response to anzu, then all of the windows will show this.

I managed to resolve this by doing this:

``` elisp
(add-hook 'anzu-mode-hook (lambda () (make-local-variable 'anzu--state)))
```

Now my mode-line is working as expected: searching in one buffer in a window won't set the anzu stuff in my mode-line for every window. I'm not sure if this is the preferred way to fix this though.

**Comment by syohex** [comment]
Could you tell me how to reproduce this issue ?
- configuration
- screenshot
- etc

**Comment by blaenk** [comment]
Ah I don't have time currently to do that, so don't worry about it since it seems like it doesn't affect many other people, feel free to close it.

Basically if I do something like:

``` elisp
(setq anzu-cons-mode-line-p nil)
```

So that now I can choose where the anzu information should show up, for example:

``` elisp
(setq-default mode-line-format '((anzu-mode (:eval (anzu--update-mode-line)))))
```

That uses the default mode-line setting format, which only draws the anzu stuff if `anzu--state` is set. Normally what happens if you don't have `anzu-cons-mode-line-p` set to nil, is that you go into search mode, then it performs the cons to the current mode line and sets `anzu--state` to non-nil, so even though `anzu--state` is global, when you switch to another window it won't render the anzu stuff because to do so it'd have to cons, and it already consed in the other window (since mode-line-format is buffer local).

When you disable consing, what happens is that it sets `anzu--state` to non-nil and calls your mode-line update function, which is great, but then if I switch to the other window/buffer, the mode-line-format will check to see if `anzu--state` is non-nil to determine whether or not to show anzu info, and it _will_ be non-nil because the other buffer set it to non-nil, and it's global.

Hope that was a bit clearer. It was a simple fix on my end, I imagine it'd be simple to just make it buffer local in this package, but I don't know if that would have unintended consequences. In the meantime the fix on my end works well.

**Comment by stardiviner** [comment]
I set custom mode-line with this:
```lisp
(:propertize (:eval (anzu--update-mode-line))
                face (:foreground "cyan" :weight 'bold
                                  :box '(:color "deep pink")))
```

and have anzu status on mode-line like in this screenshot:


![Uploading anzu--state-global.png…]()

**Comment by syohex** [comment]
Please show me all configuration about anzu.

**Comment by stardiviner** [comment]
```lisp

(setq anzu-regexp-search-commands '(vr/isearch-forward
                                    vr/isearch-backward
                                    isearch-forward-regexp
                                    isearch-backward-regexp)
      ;; anzu--mode-line-format '(:eval (anzu--update-mode-line)) ; add into my custom mode-line
      anzu--mode-line-format ""
      ;; anzu-cons-mode-line-p
      ;; anzu-input-idle-delay 0.05
      anzu-deactivate-region nil
      anzu-use-migemo nil
      anzu-cons-mode-line-p nil
      anzu-replace-to-string-separator " ⇨ "
      ;; anzu-minimum-input-length 1
      anzu-search-threshold nil ; limit of search number.
      )

(global-anzu-mode +1)
;; (anzu-mode +1)

;; (global-set-key (kbd "M-%") 'anzu-query-replace)
(global-set-key (kbd "M-%") 'anzu-query-replace-regexp)
(global-set-key (kbd "C-M-%") 'anzu-query-replace-regexp)
```

**Comment by syohex** [comment]
Thanks.

**Comment by syohex** [comment]
Is this issue resolved by following patch ?

```diff
diff --git a/anzu.el b/anzu.el
index 568fc24..601e04f 100644
--- a/anzu.el
+++ b/anzu.el
@@ -304,6 +304,7 @@
   :lighter    anzu-mode-lighter
   (if anzu-mode
       (progn
+        (set (make-local-variable 'anzu--state) nil)
         (add-hook 'isearch-update-post-hook 'anzu--update-post-hook nil t)
         (add-hook 'isearch-mode-hook 'anzu--cons-mode-line-search nil t)
         (add-hook 'isearch-mode-end-hook 'anzu--reset-mode-line nil t))
```

**Comment by stardiviner** [comment]
@syohex it does not resolve the issue.

**Comment by blaenk** [comment]
I'm pretty sure that should definitely have resolved it, it's what solved it for me. Maybe that's not the way it should be patched? Try doing it the way I did it:

``` elisp
;; in your init.el
(add-hook 'anzu-mode-hook (lambda () (make-local-variable 'anzu--state)))
```

You made sure to remove the compiled elisp bytecode file, if you had one, after making the change, right? Just trying to think of why it's not working for you.

**Comment by stardiviner** [comment]
@blaenk after recompile file, it can work. as you said.

**Comment by blaenk** [comment]
Great! There you go @syohex :)

**Comment by stardiviner** [comment]
@blaenk I think @syohex 's method is better.

**Comment by blaenk** [comment]
I agree, mine was a workaround without modifying the anzu source. I was telling you to try it in case the other one didn't work for whatever reason.

**Comment by stardiviner** [comment]
Oh, sorry, I misunderstand.

**Comment by blaenk** [comment]
No problem!

**Comment by syohex** [comment]
Thanks for checking. I have merged at #59.
